File: src/testp1/Player.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package testp1;
import java.util.ArrayList;
import java.util.Arrays;

/**
 *
 * @author usuario
 */
public class Player{
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH =10;
    private static final int HITS2LOSE = 3;
    private String name;
    private char number;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;
    private int consecutiveHits = 0;
    
    private ArrayList <Weapon> weapons;
    private ArrayList <Shield> shields;
    
    public Player (char _number, float _intelligence, float _strength){
        name = "Player #" + number;
        number=_number;
        intelligence=_intelligence;
        strength=_strength;
        weapons= new ArrayList <>();
        shields= new ArrayList <>();
    }
    
    public void resurrect(){
        health = INITIAL_HEALTH;
        resetHits();
        weapons.clear();
        shields.clear();
    }
    
    public int getRow(){
        return row;
    }
    
    public int getCol(){
        return col;
    }
    
    public int getNumber(){
        return number;
    }
    
    public void setPos(int _row, int _col){
        row=_row;
        col=_col;
    }
    
    public boolean dead(){
        if (health <= 0)
            return true;
        else
            return false;
    }
    
    public Directions move (Directions direction, ArrayList <Directions> validMoves){
        throw new UnsupportedOperationException();
    }
    
    public float attack(){
        return strength + sumWeapons();
    }
    
    public boolean defend (float receivedAttack){
        return manageHit(receivedAttack);
    }
    
    public void receiveReward(){
        throw new UnsupportedOperationException();
    }
    
    public String toString(){
        return "Nombre: " + name + "\nInteligencia: " + intelligence +
                "\nFuerza: " + strength + "\nSalud: " + health + "\nFila: " +
                row + "\nColumna: " + col + "\nGolpes Consecutivosa: " + consecutiveHits;
    }
    
    private void receiveWeapon(){
        throw new UnsupportedOperationException();
    }
    
    private void receiveShield(){
        throw new UnsupportedOperationException();
    }
    
    private Weapon newWeapon(){
        Weapon arma;
        arma = new Weapon(Dice.weaponPower(), Dice.usesLeft());
        return arma;
    }
    
    private Shield newShield(){
        Shield escudo;
        escudo = new Shield(Dice.shieldPower(), Dice.usesLeft());
        return escudo;
    }
    
    private float sumWeapons(){
        float total=0;
        for(int i=0; i<weapons.size(); i++){
            total+=weapons.get(i).attack();
        }
        
        return total;
    }
    
    private float sumShields(){
        float total=0;
        for(int i=0; i<shields.size(); i++){
            total+=shields.get(i).protect();
        }
        
        return total;
    }
    
    private float defensiveEnergy(){
        return intelligence + sumShields();
    }
    
    private boolean manageHit(float receivedAttack){
        throw new UnsupportedOperationException();
    }
    
    private void resetHits(){
        consecutiveHits=0;
    }
    
    private void gotWounded(){
        health--;
    }
    
    private void incConsecutiveHits(){
        consecutiveHits++;
    }
}

File: RUBY/player.rb
# frozen_string_literal: true

require_relative 'weapon'
require_relative 'shield'
require_relative 'dice'
require_relative 'directions'


module Irrgarten
class Player
  @@MAX_WEAPONS = 2
  @@MAX_SHIELDS = 3
  @@INITIAL_HEALTH = 10
  @@HITS2LOSE = 3

  attr_reader :number
  def initialize (num,  inteligencia, fuerza)
    @number = num
    @name = "Player ##{@number}"
    @intelligence = inteligencia
    @strength = fuerza
    @health = @@INITIAL_HEALTH
    @consecutiveHits = 0
    @weapons = []
    @shields = []
  end

  def getRow
    @row
  end

  def getCol
    @col
  end

  def getNumber
   @number
  end

  def setPos(row, col)
    @row = row
    @col = col
  end

  def resurrect
    @health=@@INITIAL_HEALTH
    resetHits
    @weapons.clear
    @shields.clear
  end

  def dead
    @health<=0
  end

  def move(direction, validMoves)
    if validMoves.size > 0 && !validMoves.include?(direction)
      firstElement=validMoves[0]
    else
      direction
    end
  end

  def attack
    @strength + sumWeapons
  end

  def defend (receivedAttack)
    manageHit(receivedAttack)
  end

  def receiveReward
    wReward = Dice.weaponsReward
    sReward = Dice.shieldsReward
    for i in (0...wReward)
      wnew = Weapon.new(Dice.weaponPower, Dice.usesLeft)
      receiveWeapon(wnew)
    end
    for i in (0...sReward)
      snew = Shield.new(Dice.shieldPower, Dice.usesLeft)
      receiveShield(snew)
    end
    @health += Dice.healthReward
  end

  def to_s
    "Nombre del Jugador: #{@name}, \nNumero: #{@number}, \nFuerza: #{@strength}, \nInteligencia: #{@intelligence},
    \nSalud: #{@health}, \nFila: #{@row}, \nColumna: #{@col}, \nGolpes Consecutivos: #{@consecutiveHits}, \nArmas: #{@weapons},
    \nEscudos: #{@shields}"
  end

  private
  def receiveWeapon(w)
    for i in (0...@weapons.size)
      if @weapons[i].discard
        @weapons.delete_at(i)
      end
    end
    if @weapons.size < @@MAX_WEAPONS
      @weapons << w
    end
  end

  def receiveShield(s)
    for i in (0...@shields.size)
      if @shields[i].discard
        @shields.delete_at(i)
      end
    end
    if @shields.size < @@MAX_SHIELDS
      @shields << s
    end
  end


  def newWeapon
    Weapon.new(Dice.weaponPower, Dice.usesLeft)
  end

  def newShield
    Shield.new(Dice.shieldPower, Dice.usesLeft)
  end

  def sumWeapons
    total=0
    for i in (0...@weapons.size)
      total += @weapons[i].attack
    end
    total
  end

  def sumShields
    total=0
    for i in (0...@shields.size)
      total+=@shields[i].protect
    end
    total
  end

  def defensiveEnergy
    @intelligence + sumShields
  end

  def manageHit(receivedAttack)
    if defensiveEnergy < receivedAttack
      gotWounded
      incConsecutiveHits
    else
      resetHits
    end
    if @consecutiveHits == @@HITS2LOSE || dead
      resetHits
      lose=true
    else
      lose=false
    end
  end

  def resetHits
    @consecutiveHits=0
  end

  def gotWounded
    @health-=1
  end

  def incConsecutiveHits
    @consecutiveHits+=1
  end

end
end